midend/has_side_effects.h

```diff
@@ -18,18 +18,38 @@ limitations under the License.
 #define MIDEND_HAS_SIDE_EFFECTS_H_
 
 #include "ir/ir.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+#include "frontends/common/resolveReferences/referenceMap.h"
 
-/* Should this be a method on IR::Expression? */
+/* Should this be a method on IR::Expression?  Maybe after the refMap/typeMap go away */
 
 class hasSideEffects : public Inspector {
+    P4::ReferenceMap    *refMap = nullptr;
+    P4::TypeMap         *typeMap = nullptr;
+
     bool result = false;
```

Comment:
Is `result` supposed to be a memoized result? If so, I'd change its type to be something like `boost::optional<bool>`, to avoid conflating a memoized `false` with a result that still needs to be computed.

Reply:
`false` means we have not yet seen a side effect anywhere we've looked.  Once we see a side effect, we set it to true and prune all future traversals.